rcl/src/rcl/validate_security_context_name.c

```diff
@@ -0,0 +1,148 @@
+}
+
+rcl_ret_t
+rcl_validate_security_context_name_with_size(
+  const char * security_context,
+  size_t security_context_length,
+  int * validation_result,
+  size_t * invalid_index)
+{
+  if (!security_context) {
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+  if (!validation_result) {
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  int t_validation_result;
+  size_t t_invalid_index;
+  rmw_ret_t ret = rmw_validate_namespace_with_size(
+    security_context, security_context_length, &t_validation_result, &t_invalid_index);
+  if (ret != RMW_RET_OK) {
+    return rcl_convert_rmw_ret_to_rcl_ret(ret);
+  }
+
+  if (t_validation_result != RMW_NAMESPACE_VALID &&
+    t_validation_result != RMW_NAMESPACE_INVALID_TOO_LONG)
+  {
+    switch (t_validation_result) {
+      case RMW_NAMESPACE_INVALID_IS_EMPTY_STRING:
+        *validation_result = RCL_SECURITY_CONTEXT_NAME_INVALID_IS_EMPTY_STRING;
+        break;
+      case RMW_NAMESPACE_INVALID_NOT_ABSOLUTE:
+        *validation_result = RCL_SECURITY_CONTEXT_NAME_INVALID_NOT_ABSOLUTE;
+        break;
+      case RMW_NAMESPACE_INVALID_ENDS_WITH_FORWARD_SLASH:
+        *validation_result = RCL_SECURITY_CONTEXT_NAME_INVALID_ENDS_WITH_FORWARD_SLASH;
+        break;
+      case RMW_NAMESPACE_INVALID_CONTAINS_UNALLOWED_CHARACTERS:
+        *validation_result = RCL_SECURITY_CONTEXT_NAME_INVALID_CONTAINS_UNALLOWED_CHARACTERS;
+        break;
+      case RMW_NAMESPACE_INVALID_CONTAINS_REPEATED_FORWARD_SLASH:
+        *validation_result = RCL_SECURITY_CONTEXT_NAME_INVALID_CONTAINS_REPEATED_FORWARD_SLASH;
+        break;
+      case RMW_NAMESPACE_INVALID_NAME_TOKEN_STARTS_WITH_NUMBER:
+        *validation_result = RCL_SECURITY_CONTEXT_NAME_INVALID_NAME_TOKEN_STARTS_WITH_NUMBER;
+        break;
+      default:
+        {
+          char default_err_msg[256];
+          // explicitly not taking return value which is number of bytes written
+          int ret = rcutils_snprintf(
+            default_err_msg, sizeof(default_err_msg),
+            "rcl_validate_security_context_name_with_size(): "
+            "unknown rmw_validate_namespace_with_size() result '%d'",
+            t_validation_result);
+          if (ret < 0) {
+            RCL_SET_ERROR_MSG(
+              "rcl_validate_security_context_name_with_size(): "
+              "rcutils_snprintf() failed while reporting an unknown validation result");
+          } else {
+            RCL_SET_ERROR_MSG(default_err_msg);
+          }
+        }
+        return RCL_RET_ERROR;
+    }
+    if (invalid_index) {
+      *invalid_index = t_invalid_index;
+    }
+    return RCL_RET_OK;
+  }
+
+  // security_context might be longer that namespace length, check false positives and correct
+  if (t_validation_result == RMW_NAMESPACE_INVALID_TOO_LONG &&
+    security_context_length <= RCL_SECURITY_CONTEXT_NAME_MAX_LENGTH)
+  {
+    *validation_result = RCL_SECURITY_CONTEXT_NAME_VALID;
+    return RCL_RET_OK;
+  }
```

Comment:
@ivanpauno Hmm I though I had brought up this. What if it is too long and it exceeds `RCL_SECURITY_CONTEXT_NAME_MAX_LENGTH`?

Reply:
See https://github.com/ros2/rcl/pull/515#discussion_r395865974.